packages/server-wallet/src/protocols/defunder.ts

```diff
@@ -0,0 +1,104 @@
+import {unreachable} from '@statechannels/wallet-core';
+import {Transaction} from 'objection';
+import {Logger} from 'pino';
+
+import {ChainServiceInterface} from '../chain-service';
+import {AdjudicatorStatusModel} from '../models/adjudicator-status';
+import {ChainServiceRequest} from '../models/chain-service-request';
+import {Channel} from '../models/channel';
+import {LedgerRequest} from '../models/ledger-request';
+import {Store} from '../wallet/store';
+
+export type DefunderResult = {isChannelDefunded: boolean; didSubmitTransaction: boolean};
+
+export class Defunder {
+  constructor(
+    private store: Store,
+    private chainService: ChainServiceInterface,
+    private logger: Logger,
+    private timingMetrics = false
+  ) {}
+
+  public static create(
+    store: Store,
+    chainService: ChainServiceInterface,
+    logger: Logger,
+    timingMetrics = false
+  ): Defunder {
+    return new Defunder(store, chainService, logger, timingMetrics);
+  }
+
+  public async isChannelDefunded(channel: Channel, tx: Transaction): Promise<DefunderResult> {
+    const {protocolState: ps} = channel;
+
+    switch (ps.fundingStrategy) {
+      case 'Direct':
+        return await this.directDefunder(channel, tx);
+      case 'Ledger':
+        return this.ledgerDefunder(channel, tx);
+      case 'Unknown':
+      case 'Fake':
+        return {isChannelDefunded: true, didSubmitTransaction: false};
+      case 'Virtual':
+        throw new Error('Virtual channel defunding is not implemented');
+      default:
+        unreachable(ps.fundingStrategy);
+    }
+  }
+
+  private async directDefunder(channel: Channel, tx: Transaction): Promise<DefunderResult> {
+    if (!channel.isPartlyDirectFunded) {
+      return {isChannelDefunded: true, didSubmitTransaction: false};
+    }
+
+    const hasValidRequest = ['withdraw', 'pushOutcome']
+      .map(requestType =>
+        channel.chainServiceRequests.find(csr => csr.request === requestType)?.isValid()
+      )
+      .some(val => val);
+    if (hasValidRequest) return {isChannelDefunded: false, didSubmitTransaction: false};
+
+    const adjudicatorStatus = await AdjudicatorStatusModel.getAdjudicatorStatus(
+      tx,
+      channel.channelId
+    );
+    let didSubmitTransaction = false;
+    if (adjudicatorStatus.channelMode !== 'Finalized' && channel.hasConclusionProof) {
+      await ChainServiceRequest.insertOrUpdate(channel.channelId, 'withdraw', tx);
+
+      // supported is defined (if the wallet is functioning correctly), but the compiler is not aware of that
```

Comment:
nFor cases like this (where we always expect something to be defined if the wallet is working correctly), I've often added an `if` statement and then log a warning or error.

 I'm not sure if that's the right solution but I feel like this is a pretty common case so it might make sense for us to adopt a common pattern?

Reply:
This is a good point. Addressed in b1f22f2b9

I prefer throwing an error. In this case, if `channel.supported` is undefined `chainService.concludeAndWithdraw` will not succeed.